From 5a869b04a30cd16237ee3dfa996d77bfc3574cb9 Mon Sep 17 00:00:00 2001
From: Samuel Chiang <sachiang@amazon.com>
Date: Mon, 2 Feb 2026 10:00:38 -0800
Subject: [PATCH] [Cherry-pick 2024] Offer P521 for signature_algorithms in
 client Hello (#2975)

(cherry picked from commit 0beb210b02b48d74444002393812a1a07a3b63b9).

This change is not in the FIPS boundary and does not impact the
integrity hash.

By submitting this pull request, I confirm that my contribution is made
under the terms of the Apache 2.0 license and the ISC license.

Co-authored-by: Will Childs-Klein <childw@amazon.com>
---
 ssl/ssl_test.cc | 40 ++++++++++++++++++++++++++++++++++++----
 1 file changed, 36 insertions(+), 4 deletions(-)

diff --git a/ssl/ssl_test.cc b/ssl/ssl_test.cc
index 0cdec36ac..e2ee69afc 100644
--- a/ssl/ssl_test.cc
+++ b/ssl/ssl_test.cc
@@ -4747,8 +4747,27 @@ TEST(SSLTest, ClientHello) {
         0x02, 0x01, 0x00, 0x00, 0x23, 0x00, 0x00, 0x00, 0x0d, 0x00, 0x16, 0x00,
         0x14, 0x04, 0x03, 0x08, 0x04, 0x04, 0x01, 0x05, 0x03, 0x08, 0x05, 0x05,
         0x01, 0x06, 0x03, 0x08, 0x06, 0x06, 0x01, 0x02, 0x01}},
-      // TODO(davidben): Add a change detector for TLS 1.3 once the spec and our
-      // implementation has settled enough that it won't change.
+      {TLS1_3_VERSION,
+       {0x16, 0x03, 0x01, 0x00, 0xeb, 0x01, 0x00, 0x00, 0xe7, 0x03, 0x03, 0x00,
+        0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+        0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+        0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+        0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+        0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+        0x00, 0x00, 0x00, 0x00, 0x00, 0x28, 0x13, 0x01, 0x13, 0x02, 0x13, 0x03,
+        0xcc, 0xa9, 0xcc, 0xa8, 0xc0, 0x2b, 0xc0, 0x2f, 0xc0, 0x2c, 0xc0, 0x30,
+        0xc0, 0x09, 0xc0, 0x13, 0xc0, 0x27, 0xc0, 0x0a, 0xc0, 0x14, 0xc0, 0x28,
+        0x00, 0x9c, 0x00, 0x9d, 0x00, 0x2f, 0x00, 0x3c, 0x00, 0x35, 0x01, 0x00,
+        0x00, 0x76, 0x00, 0x17, 0x00, 0x00, 0xff, 0x01, 0x00, 0x01, 0x00, 0x00,
+        0x0a, 0x00, 0x08, 0x00, 0x06, 0x00, 0x1d, 0x00, 0x17, 0x00, 0x18, 0x00,
+        0x0b, 0x00, 0x02, 0x01, 0x00, 0x00, 0x23, 0x00, 0x00, 0x00, 0x0d, 0x00,
+        0x16, 0x00, 0x14, 0x04, 0x03, 0x08, 0x04, 0x04, 0x01, 0x05, 0x03, 0x08,
+        0x05, 0x05, 0x01, 0x06, 0x03, 0x08, 0x06, 0x06, 0x01, 0x02, 0x01, 0x00,
+        0x33, 0x00, 0x26, 0x00, 0x24, 0x00, 0x1d, 0x00, 0x20, 0x00, 0x00, 0x00,
+        0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+        0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+        0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x2d, 0x00, 0x02, 0x01, 0x01, 0x00,
+        0x2b, 0x00, 0x09, 0x08, 0x03, 0x04, 0x03, 0x03, 0x03, 0x02, 0x03, 0x01}},
   };
 
   for (const auto &t : kTests) {
@@ -4771,8 +4790,21 @@ TEST(SSLTest, ClientHello) {
     constexpr size_t kRandomOffset = 1 + 2 + 2 +  // record header
                                      1 + 3 +      // handshake message header
                                      2;           // client_version
-    ASSERT_GE(client_hello.size(), kRandomOffset + SSL3_RANDOM_SIZE);
-    OPENSSL_memset(client_hello.data() + kRandomOffset, 0, SSL3_RANDOM_SIZE);
+
+    int pre = client_hello.size();
+    if (t.max_version == TLS1_3_VERSION) {
+      ASSERT_GE(client_hello.size(),
+                kRandomOffset + SSL3_RANDOM_SIZE + 1 + SSL3_SESSION_ID_SIZE);
+      OPENSSL_memset(client_hello.data() + kRandomOffset, 0,
+                     SSL3_RANDOM_SIZE + 1 + SSL3_SESSION_ID_SIZE);
+      // Jump to key share extension and zero out the key
+      OPENSSL_memset(client_hello.data() + 189, 0, 32);
+    } else {
+      ASSERT_GE(client_hello.size(), kRandomOffset + SSL3_RANDOM_SIZE);
+      OPENSSL_memset(client_hello.data() + kRandomOffset, 0, SSL3_RANDOM_SIZE);
+    }
+    int post = client_hello.size();
+    ASSERT_EQ(pre, post);
 
     if (client_hello != t.expected) {
       ADD_FAILURE() << "ClientHellos did not match.";
